presto-base-arrow-flight/src/main/java/com/facebook/plugin/arrow/ArrowPageUtils.java

```diff
@@ -0,0 +1,847 @@
+            return buildBlockFromFloat8Vector((Float8Vector) vector, type);
+        }
+        else if (vector instanceof VarCharVector) {
+            if (type instanceof CharType) {
+                return buildCharTypeBlockFromVarcharVector((VarCharVector) vector, type);
+            }
+            else if (type instanceof TimeType) {
+                return buildTimeTypeBlockFromVarcharVector((VarCharVector) vector, type);
+            }
+            else {
+                return buildBlockFromVarCharVector((VarCharVector) vector, type);
+            }
+        }
+        else if (vector instanceof VarBinaryVector) {
+            return buildBlockFromVarBinaryVector((VarBinaryVector) vector, type);
+        }
+        else if (vector instanceof DateDayVector) {
+            return buildBlockFromDateDayVector((DateDayVector) vector, type);
+        }
+        else if (vector instanceof DateMilliVector) {
+            return buildBlockFromDateMilliVector((DateMilliVector) vector, type);
+        }
+        else if (vector instanceof TimeMilliVector) {
+            return buildBlockFromTimeMilliVector((TimeMilliVector) vector, type);
+        }
+        else if (vector instanceof TimeSecVector) {
+            return buildBlockFromTimeSecVector((TimeSecVector) vector, type);
+        }
+        else if (vector instanceof TimeStampSecVector) {
+            return buildBlockFromTimeStampSecVector((TimeStampSecVector) vector, type);
+        }
+        else if (vector instanceof TimeMicroVector) {
+            return buildBlockFromTimeMicroVector((TimeMicroVector) vector, type);
+        }
+        else if (vector instanceof TimeStampMilliTZVector) {
+            return buildBlockFromTimeMilliTZVector((TimeStampMilliTZVector) vector, type);
+        }
+        else if (vector instanceof ListVector) {
+            return buildBlockFromListVector((ListVector) vector, type);
+        }
+
+        throw new UnsupportedOperationException("Unsupported vector type: " + vector.getClass().getSimpleName());
+    }
+
+    public static Block buildBlockFromEncodedVector(IntVector encodedVector, VarCharVector dictionary)
+    {
+        // Ensure the dictionary vector is valid
+        if (dictionary == null || encodedVector == null) {
+            throw new IllegalArgumentException("Both encodedVector and dictionary must be non-null.");
+        }
+
+        // Create a BlockBuilder for VARCHAR
+        BlockBuilder builder = VarcharType.VARCHAR.createBlockBuilder(null, encodedVector.getValueCount());
+
+        // Iterate through the encoded vector and retrieve values from the dictionary
+        for (int i = 0; i < encodedVector.getValueCount(); i++) {
+            if (encodedVector.isNull(i)) {
+                builder.appendNull(); // Append null if the index is null
+            }
+            else {
+                int dictionaryIndex = encodedVector.get(i);
+                if (dictionary.isNull(dictionaryIndex)) {
```

Comment:
can't we use `Dictionary` and `DictionaryEncoder.decode` https://arrow.apache.org/docs/java/ipc.html